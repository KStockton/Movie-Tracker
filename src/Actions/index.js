export const addTopMovies = topMovies => ({
    type: "ADD_TOP_MOVIES",
    topMovies
})