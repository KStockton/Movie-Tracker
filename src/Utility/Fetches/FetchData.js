export const fetchData = async url => {
  let response = await fetch(url);
  let resolved = await response.json();
  return resolved
};

