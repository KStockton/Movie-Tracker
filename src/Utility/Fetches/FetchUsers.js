const settings = (method, body) => ({
    method,
    headers: {'Content-Type': 'application/json'},
    body: JSON.stringify(body)
})
const fetchUser = async (url) => {
    const response = await fetch('https://movie-tracker-fe.herokuapp.com/api/')
}