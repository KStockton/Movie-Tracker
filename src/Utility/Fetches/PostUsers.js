const fetchConfig = (method, body) => ({
  method,
  headers: { "Content-Type": "application/json" },
  body: JSON.stringify(body)
});
const postUsers = async (url, method, body) => {
  const response = await fetch(
    `http://localhost:3000/api/${url}`,
    fetchConfig(method, body)
  );
  const result = await response.json();
  return result;
};
export { postUsers, fetchConfig };
