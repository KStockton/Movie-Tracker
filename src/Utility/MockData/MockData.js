
export const uncleanMovies =
[
  {
      "vote_count": 2045,
      "id": 19404,
      "video": false,
      "vote_average": 9,
      "title": "Dilwale Dulhania Le Jayenge",
      "poster_path": "/uC6TTUhPpQCmgldGyYveKRAu8JN.jpg",
      "original_language": "hi",
      "original_title": "दिलवाले दुल्हनिया ले जायेंगे",
      "genre_ids": [
          35,
          18,
          10749
      ],
      "backdrop_path": "/nl79FQ8xWZkhL3rDr1v2RFFR6J0.jpg",
      "adult": false,
      "overview": "Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.",
      "release_date": "1995-10-20"
  },
  {
      "vote_count": 12962,
      "id": 278,
      "video": false,
      "vote_average": 8.7,
      "title": "The Shawshank Redemption",
      "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      "original_language": "en",
      "original_title": "The Shawshank Redemption",
      "genre_ids": [
          18,
          80
      ],
      "backdrop_path": "/j9XKiZrVeViAixVRzCta7h1VU9W.jpg",
      "adult": false,
      "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      "release_date": "1994-09-23"
  }]

  export const cleanMovies = 
  [{
     id: 19404,
     title: "Dilwale Dulhania Le Jayenge",
     backdrop_path: "/nl79FQ8xWZkhL3rDr1v2RFFR6J0.jpg",
     overview: "Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancé. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.",
     poster_path: "/uC6TTUhPpQCmgldGyYveKRAu8JN.jpg",
     vote_average: 9,
     release_date: "1995-10-20",
     favorited: false
  },
    {
      id: 278,
      title:  "The Shawshank Redemption",
      backdrop_path: "/j9XKiZrVeViAixVRzCta7h1VU9W.jpg",
      overview: "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
      poster_path: "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
      vote_average: 8.7,
      release_date: "1994-09-23",
      favorited: false
  }]

  export const movieProp = {
    state: {
      movie: {
        id: 2341,
        title: "HellBoy",
        backdrop_path: "/5BkSkNtfrnT",
        favorited: false,
        overview: "walk on earth",
        poster_path: "/bk8LyaMqUtaQ9hUShuvFznQYQKR.jpg",
        release_date: "2019-04-12"
      }
    }
  }
