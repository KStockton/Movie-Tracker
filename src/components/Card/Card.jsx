import React from "react";
import noFav from "../../Images/noFav.svg";
import fav from "../../Images/fav.svg";

const Card = ({ movie }) => {
  console.log(movie);
  const {
    id,
    overview,
    popularity,
    favorite,
    posterpath,
    image,
    title
  } = movie;
  console.log(movie);
  const imagesrc = `https://image.tmdb.org/t/p/w500${posterpath}`;

  return (
    <div className="card-wrapper">
      <div
        className="card-background"
        style={{
          backgroundImage: `url(https://image.tmdb.org/t/p/w500${image})`
        }}
      >
        <article className="card-info">
          <h4 className="card-title">{title}</h4>
          <span>{popularity}</span>
          <button>More Info</button>
        </article>
      </div>

      {/* {popularity && <p className="Card-info">Overview {popularity}</p>}
    <img className="Card-no-favorite" src={favorite ? fav : noFav} alt="favorite"/>
    <img src={imagesrc} alt="poster"/>
    <button>Favorite</button> */}
    </div>
  );
};
export default Card;
