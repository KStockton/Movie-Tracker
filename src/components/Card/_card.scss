.card-wrapper {
  height: 20em;
  margin: 0em 0.5em;
  position: relative;
}

.card-background {
  height: 18.2em;
  width: 12.1em;
  outline: grey 0.1em solid;
  background-size: contain;
  transition: transform 0.3s;
  background-repeat: no-repeat;
  // border: grey solid 2px;
}
.card-background:hover {
  transform: scale(1.3);
}
.card-wrapper:hover {
  z-index: 50;
  transform: rotate(0.01deg);
  opacity: 1;
}

.card-info {
  opacity: 0;
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
}
.card-info:hover {
  opacity: 1;
  width: 100%;
  height: 100%;
  transform: rotate(0.01deg);

}

.card-title {
  text-overflow: ellipsis;
  white-space: nowrap;
  overflow: hidden;
  font-size: 1em;
  font-weight: normal;
}

.favorite-icon {
pointer-events: none;

}

.card-button {
  background-color: transparent;
  font-size: 1.5em;
  color: $user-text-color-main;
  text-decoration: none;
  transition: color 0.4s;
  text-shadow:
   -1px -1px 0 #000,  
    1px -1px 0 #000,
    -1px 1px 0 #000,
     1px 1px 0 #000;
}
.card-button:hover {
  cursor: pointer;
}
.card-button-container {
  display: flex;
  justify-content: space-between;
  align-items: flex-end;
  padding: 1em; 
  height: 100%;
  flex-direction: column;
}

.more-info-icon {
  color: white;
}
.card-rating {
  color: rgb(44, 175, 44);
  font-weight: bold;
  font-size: 0.3em;
}
.rating {
  font-size: 0.3em;
}

button { 
  outline: none;
  border: none;
}