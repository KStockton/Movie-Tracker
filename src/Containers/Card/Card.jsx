import React, { Component } from "react";
import { postUsers } from "../../Utility/Fetches/PostUsers";
import { NavLink } from "react-router-dom";
import { connect } from "react-redux";
import { setUserFavorites } from "../../Actions";
import { library } from "@fortawesome/fontawesome-svg-core";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faHeart,
  faArrowAltCircleRight
} from "@fortawesome/free-solid-svg-icons";
import PropTypes from "prop-types";
library.add(faArrowAltCircleRight, faHeart);

export class Card extends React.Component {
  handleFavorite = async () => {
    const { movie, user } = this.props;
    const {
      id,
      overview,
      vote_average,
      release_date,
      poster_path,
      title
    } = movie;
    const userFavInfo = {
      movie_id: id,
      user_id: user.id,
      title: title,
      poster_path: poster_path,
      vote_average: vote_average,
      release_date: release_date,
      overview: overview
    };

    const uniqueMovie = this.props.user.favorites.filter(
      movie => movie.movie_id === userFavInfo.movie_id
    );

    if (this.props.user !== undefined && uniqueMovie.length === 0) {
      const favPath = "users/favorites/new";
      await postUsers(favPath, "POST", userFavInfo);
      this.setFavorites(user);
    } else {
      const path = `users/${this.props.user.id}/favorites/${
        userFavInfo.movie_id
      }`;
      await postUsers(path, "DELETE", {
        id: this.props.user.id,
        movie_id: userFavInfo.movie_id
      });
      this.setFavorites(user);
    }
  };

  setFavorites = async user => {
    const checkFavePath = `users/${user.id}/favorites`;
    let faves = await postUsers(checkFavePath, "GET");
    this.props.setUserFavorites(faves.data);
  };

  checkFavorite = () => {
    const favoriteIDs = this.getFavoriteIDs();
    if (favoriteIDs.includes(this.props.movie.id)) {
      const displayMovie = { ...this.props.movie, favorited: true };
      return displayMovie;
    } else {
      const displayMovie = { ...this.props.movie };
      return displayMovie;
    }
  };

  getFavoriteIDs = () => {
    return this.props.user.favorites !== undefined
      ? this.props.user.favorites.map(movie => movie.movie_id)
      : [];
  };

  render() {
    const displayMovie = this.checkFavorite();
    const { name } = this.props.user;
    const { poster_path, favorited, id } = displayMovie;
    return (
      <div className="card-wrapper">
        <div
          className="card-background"
          style={{
            backgroundImage: `url(https://image.tmdb.org/t/p/w500${poster_path})`
          }}
        >
          <article className="card-info">
            <section className="card-button-container">
              {name === undefined ? (
                <NavLink
                  to="/login"
                  className="card-button"
                  style={{ color: favorited ? "red" : "grey" }}
                >
                  <FontAwesomeIcon className="favorite-icon" icon="heart" />
                </NavLink>
              ) : (
                <button
                  className="card-button"
                  onClick={() => this.handleFavorite()}
                  style={{ color: favorited ? "red" : "grey" }}
                >
                  <FontAwesomeIcon icon="heart" className="favorite-icon" />
                </button>
              )}
              <NavLink
                to={{
                  pathname: `/Movie/${id}`,
                  state: { movie: this.props.movie }
                }}
                className="card-button"
              >
                <FontAwesomeIcon
                  className="more-info-icon"
                  icon="arrow-alt-circle-right"
                />
              </NavLink>
            </section>
          </article>
        </div>
      </div>
    );
  }
}
export const mapStateToProps = state => ({
  user: state.user,
  topMovies: state.topMovies
});
export const mapDispatchToProps = dispatch => ({
  setUserFavorites: user => dispatch(setUserFavorites(user))
});

Card.propTypes = {
  movie: PropTypes.object,
  user: PropTypes.object,
  topMovies: PropTypes.array,
  setUserFavorites: PropTypes.func
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Card);
