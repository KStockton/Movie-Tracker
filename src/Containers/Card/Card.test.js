
import React from "react";
import { Card, mapStateToProps, mapDispatchToProps } from "./Card";
import { shallow } from "enzyme";
import { postUsers } from '../../Utility/Fetches/PostUsers';
import { setUserFavorites } from '../../Actions/index';
import * as MD from '../../Utility/MockData/MockData'

const setFavoritesmock =jest.fn()


 describe("Card", () => {
  let wrapper;
  jest.mock('../../Utility/Fetches/PostUsers');

   const mockUser = { favorites: [{ id: 1, movie_id: 1 }] };
  const mockMovie = {   
    id: 2,
    movie_id: 2,
    overview: "overview-text",
    vote_average: 1,
    release_date: "2019",
    poster_path: "path",
    title: "title-text"
  };

  beforeEach(() => {
    wrapper = shallow(<Card user={MD.mockUserWithFav} movie={MD.cleanMovies} />);
  }); 

   it("Card component should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call dispatch when using addTopMovies props function', () => {
    const mockDispatch = jest.fn()
    const actionToDispatch = setUserFavorites(MD.mockUserWithFav)

    const mappedProps = mapDispatchToProps(mockDispatch)
    mappedProps.setUserFavorites(MD.mockUserWithFav)
    
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
  })

  it('should return a props object with a user', () => {
    const mockState = {
      user: {
        id: 55,
        name: "nim",
        password: 'nimisawesome',
        email: "nim@sum.com"
      }
    }

    const expected = {
      user: mockState.user
    } 

    const mappedProps = mapStateToProps(mockState)

    expect(mappedProps).toEqual(expected)
  })

  it('should return a props object with a movies array', () => {
    const mockState = { topMovies: [{ 'title': "Space Jam" }]}
    const expected = {
      topMovies: mockState.topMovies
    }
    const mappedProps = mapStateToProps(mockState)
    expect(mappedProps).toEqual(expected)
  })

});   


   describe("setFavorites", () => {
    let wrapper
    beforeEach(() => {
      wrapper = shallow(<Card user={MD.mockUserWithFav} movie={MD.cleanMovies} />);
    }); 

    it.skip("should return displayMovie when checkFavorite is called", () => {
    wrapper.instance().checkFavorite()
    let mockgetFavoriteIDs = [{ id:12323 }, { id:342234 }]
    let mockMovieId = 12323
    
      expect(mockgetFavoriteIDs).toHaveBeenCalled();
    });
  }); 
    
      describe("handleFavorite", () => {
        const mockHandleFavorite = jest.fn()
      let wrapper
      beforeEach(() => {
        wrapper = shallow(<Card user={MD.mockUserWithFav} movie={MD.cleanMovies} handleFavorite={mockHandleFavorite}/>);
      }); 

      it('should call handle favorite when button is clicked ', () => {
        wrapper.find('#but').childAt(0).simulate('click', () => {
          expect(mockHandleFavorite).toHaveBeenCalled()
        })
      })
 });
      