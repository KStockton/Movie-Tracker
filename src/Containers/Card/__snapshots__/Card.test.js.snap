// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Card component should match the snapshot 1`] = `
<div
  className="card-wrapper"
>
  <div
    className="card-background"
    style={
      Object {
        "backgroundImage": "url(https://image.tmdb.org/t/p/w500undefined)",
      }
    }
  >
    <article
      className="card-info"
    >
      <section
        className="card-button-container"
      >
        <button
          className="card-button"
          id="but"
          onClick={[Function]}
          style={
            Object {
              "color": "red",
            }
          }
        >
          <FontAwesomeIcon
            border={false}
            className="favorite-icon"
            fixedWidth={false}
            flip={null}
            icon="heart"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </button>
        <NavLink
          className="card-button"
          to={
            Object {
              "pathname": "/Movie/undefined",
              "state": Object {
                "movie": Array [
                  Object {
                    "backdrop_path": "/nl79FQ8xWZkhL3rDr1v2RFFR6J0.jpg",
                    "favorited": false,
                    "id": 19404,
                    "overview": "Raj is a rich, carefree, happy-go-lucky second generation NRI. Simran is the daughter of Chaudhary Baldev Singh, who in spite of being an NRI is very strict about adherence to Indian values. Simran has left for India to be married to her childhood fiancÃ©. Raj leaves for India with a mission at his hands, to claim his lady love under the noses of her whole family. Thus begins a saga.",
                    "poster_path": "/uC6TTUhPpQCmgldGyYveKRAu8JN.jpg",
                    "release_date": "1995-10-20",
                    "title": "Dilwale Dulhania Le Jayenge",
                    "vote_average": 9,
                  },
                  Object {
                    "backdrop_path": "/j9XKiZrVeViAixVRzCta7h1VU9W.jpg",
                    "favorited": false,
                    "id": 278,
                    "overview": "Framed in the 1940s for the double murder of his wife and her lover, upstanding banker Andy Dufresne begins a new life at the Shawshank prison, where he puts his accounting skills to work for an amoral warden. During his long stretch in prison, Dufresne comes to be admired by the other inmates -- including an older prisoner named Red -- for his integrity and unquenchable sense of hope.",
                    "poster_path": "/9O7gLzmreU0nGkIB6K3BsJbzvNv.jpg",
                    "release_date": "1994-09-23",
                    "title": "The Shawshank Redemption",
                    "vote_average": 8.7,
                  },
                ],
              },
            }
          }
        >
          <FontAwesomeIcon
            border={false}
            className="more-info-icon"
            fixedWidth={false}
            flip={null}
            icon="arrow-alt-circle-right"
            inverse={false}
            listItem={false}
            mask={null}
            pull={null}
            pulse={false}
            rotation={null}
            size={null}
            spin={false}
            symbol={false}
            title=""
            transform={null}
          />
        </NavLink>
      </section>
    </article>
  </div>
</div>
`;
